File: PastYears/Practice_Algorithm/116. Populating Next Right Pointers in Each Node.cpp
class Solution {
public:
    Node* connect(Node* root) {
        Node *move = root;
        while(move != NULL){
            Node *level = move;
            /*this loop checks for specific level*/
            while(level != NULL && level->left != NULL && level->right != NULL){
                /*First connection for children has the same parent*/
                level->left->next = level->right;
                /*Second connnection for children has different parent but their parent has same level*/
                if(level->next!=NULL){
                    level->right->next = level->next->left;
                }
                level = level->next;
            }
            move = move->left;
        }
        return root;
    }
};

class Solution {
public:
    Node* connect(Node* root) {
        if(root == NULL) return root;
        queue<Node*>q;
        q.push(root);
        while(!q.empty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                Node* top = q.front();
                q.pop();
                if(i == size-1) top->next = NULL;
                else top->next = q.front();
                if(top->left != NULL) q.push(top->left);
                if(top->right) q.push(top->right);
            }
        }
        return root;
    }
};

class Solution {
public:
    Node* connect(Node* root) {
        Node* level = root, *move = root;
        if(root == NULL) return NULL;
        
        while(level->left){
            move = level;
            while(move){
                if(move->left){
                    move->left->next = move->right;
                }
                if(move->next)
                    move->right->next = move->next->left;
                move = move->next;
            }
            level = level->left;   
        }
        return root;
    }
};

class Solution {
public:
    Node* connect(Node* root) {
        Node* level = root, *move = root;
        if(root == NULL) return NULL;
        
        while(level->left){
            move = level;
            while(move){
                move->left->next = move->right;
                if(move->next)
                    move->right->next = move->next->left;
                move = move->next;
            }
            level = level->left;   
        }
        return root;
    }
};

File: PastYears/Practice_Algorithm/116. Populating Next Right Pointers in Each Node.java
class Solution {
    public Node connect(Node root) {
        if(root == null) return root;
        Queue<Node>q = new LinkedList<>();
        q.add(root);
        while(q.size()!=0){
            int size = q.size();
            for(int i = 0; i < size; i++){
                Node top = q.poll();
                if(i == size-1) top.next = null;
                else top.next = q.peek();
                if(top.left != null) q.add(top.left);
                if(top.right != null) q.add(top.right);
            }
        }
        return root;
    }
}

class Solution {
    public Node connect(Node root) {
        Node level = root, move= root;
        while(level !=null && level.left != null){
            move = level;
            while(move != null){
                move.left.next = move.right;
                if(move.next != null) move.right.next = move.next.left;
                move = move.next;
            }
            level = level.left;
        }
        return root;
    }
}
